### PR TITLE
Issue73

### PR DESCRIPTION
**Pull request by glandre**
Hi @sebbrudzinski,

I've finished the implementation of issue #73. However, I think we should decide some points:

1. Log messages: In addition to display messages and images, I also implemented log messages. Is that ok?

1. "Working locally" image: I'm currently using a 'cloud off' image, please check the (disconnected.png in icons directory). Alternatively, we can simply display nothing in such case. What do you think? Even so, I think it's nice to keep the _tooltip_, when the user put the mouse on the corner, the tooltip "Working locally" appears.

1. Cloud Support: Since we are planning to provide support to other services than Dropbox, I think we should design the code to be generic. Because of that, I replaced the "Dropbox" `ApplicationLogger` name to "Cloud Support". And also, the classes I created in the `remote` package also are based on "Cloud". Please, tell me if you agree with it. The way I implemented `CloudStatus` class, when we implement support for, let's say, Google Drive, we simply add a GD_CONNECTED status, and it's respectively image, and that's it. Please, see if you have any suggestions to the way I implemented it.

1. Images from Dropbox and Cloud-off I used are licensed as CC. How about we add references to the authros in the OLS license file?

### CODE REVIEW

**Comment by sebbrudzinski** [comment]
@glandre Thanks, it looks really good!

Concerning the points you have raised:
1. That's OK in my opinion. In addition I don't think we currently make enough use of the log window, so any new logs there are welcome.
2. It's OK to display the icon even if we are not connected to Dropbox. It will be yet another reminder that this application has got cloud support.
3. Perfectly fine with me.
4. License file is probably not a good place for attribution in the CC license. The attribution must happen in the application itself (maybe about page?). I'll give it some thought. Moreover, we need to know which CC license was used, a link to the original work, the title and the authors.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java [hunk 1]
```diff
@@ -326,4 +336,35 @@ private CompletionProvider createCompletionProvider() {
...
+        // Check Dropbox connection
+        DbxClient client = DbxUtil.getDbxClient();
+        if (client != null) {
+            String userToken = client.getAccessToken();
+            if (userToken != null && !userToken.isEmpty()) {
+                try {
+                    info = client.getAccountInfo();
+                    isConnected = true;
+                } catch (DbxException ex) {
+                    Exceptions.printStackTrace(ex);
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: We shouldn't print the stacktrace in this case. If I have my application conected with Dropbox and I start the application. but I'm not conected to the Internet, I get the stacktrace in a pop up window right at thr startup which is not the best.

**Comment by sebbrudzinski** [comment]
There are also conflicts on this branch that must be resolved before merging.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java (hunk 1 above)
Comment (glandre) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: OK. I'll correct it.

Comment (glandre) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: Corrected!

**Comment by glandre** [comment]
@sebbrudzinski,
I solved the conflicts, please check if it's OK now.

**Comment by glandre** [comment]
I noticed a recent change by LordFarin (I tried to mention him, but I couldn't), which also uses a singleton to ApplicationSettings.
I'm not sure whether is better to declare the single instance as public or private with a getter. For sure they will provide the same result, but for maintenance reasons I feel more comfortable by using the getter...

Anyway, I just want to keep it in one standard form. So, if this is the case, I change CloudStatus' single instance to public.

What do you think?

**Comment by glandre** [comment]
Here is the references to put in the project page:
[referece-links-for-web.txt](https://github.com/sebbrudzinski/Open-LaTeX-Studio/files/189822/referece-links-for-web.txt)

**Comment by sebbrudzinski** [comment]
@glandre Thanks for the image references, resolving conflicts and fixes! I'll do some testing and review soon. 

The reason we introduced the singleton pattern in the ApplicationSettings was to reduce disk writes/reads, Knowing that all classes use the same instance of the class, we can do a single write to the settings file at save and store that value in the instance. With each read, we can just return the stored value in the instance, without the need for read from settings file on disk, since any changes to the settings would go through that single instance.

At this very point, I'd say there's no strong need for the singleton pattern in the CloudStatus but it does not hurt either. Moreover, having singleton there can come handy when we extend cloud support. 

Concerning public getter/instance - if your singleton is exposed as a public field, any other class can assign a new instance to that exposed field. Having the field private and exposing a getter ensures that someone doesn't just re-instantiate the singleton at some point.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java [hunk 2]
```diff
@@ -60,6 +65,9 @@
     private DbxState dbxState;
 
     private AutoCompletion autoCompletion = null;
+    private static final ApplicationLogger LOGGER = new ApplicationLogger("Cloud Support");
+    private static final int AUTO_COMPLETE_DELAY = 700;
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: Remove this constant

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+     */
+    public static String getStatusMessage(int status) {
+        if (status >= 0 && status <= STATUS_MESSAGES.length) {
+            return STATUS_MESSAGES[status];
+        }
+        
+        return null;
+    }
+
+    protected CloudStatus() {
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java]: The constructor should be private if this is supposed to be singleton

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+public class CloudStatus {
+
+    private int status = STATUS_DISCONNECTED;
+
+    private final JLabel label;
+
+    /**
+     * Not connected with any cloud service provider
+     */
+    public static final int STATUS_DISCONNECTED = 0;
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java]: The statuses should probably be an enum.

**Comment by sebbrudzinski** [comment]
@glandre I tested this and it works as expected, good job! I've left three minor comments - once they are fixed we can merge this.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        IMG_CONNECTING = new ImageIcon(
+                CloudStatus.class.getResource(
+                        "/latexstudio/editor/resources/icons/connecting.gif"));
+        IMG_DISCONNECTED = new ImageIcon(
+                CloudStatus.class.getResource(
+                        "/latexstudio/editor/resources/icons/disconnected.png"));
+
+        STATUS_MESSAGES = new String[3];
+
+        STATUS_MESSAGES[STATUS_DISCONNECTED] = "Working locally";
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java]: Moreover, those messages can be moved to an enum class as well.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java (hunk 4 above)
Comment (glandre) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java]: OK.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java (hunk 5 above)
Comment (glandre) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java]: OK.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java (hunk 3 above)
Comment (glandre) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/remote/CloudStatus.java]: OK.

**Pull request merged**